include/envoy/filesystem/filesystem.h

```diff
@@ -35,32 +117,36 @@ class File {
   virtual void flush() PURE;
 };
 
-typedef std::shared_ptr<File> FileSharedPtr;
+typedef std::shared_ptr<StatsFile> StatsFileSharedPtr;
 
-/**
- * Abstraction for a file watcher.
- */
-class Watcher {
+class StatsInstance : public Instance {
 public:
-  typedef std::function<void(uint32_t events)> OnChangedCb;
-
-  struct Events {
-    static const uint32_t MovedTo = 0x1;
-  };
+  virtual ~StatsInstance() {}
 
-  virtual ~Watcher() {}
+  /**
+   * Creates a file, overriding the flush-interval set in the class.
+   *
+   * @param path The path of the file to open.
+   * @param dispatcher The dispatcher used for set up timers to run flush().
+   * @param lock The lock.
+   * @param file_flush_interval_msec Number of milliseconds to delay before flushing.
+   */
+  virtual StatsFileSharedPtr
+  createStatsFile(const std::string& path, Event::Dispatcher& dispatcher,
```

Comment:
Yeah I'm doubling down on not calling this a Stats file. Basically there are 2 Envoy semantics being added to raw OS files here: stats  and auto-flushing.

I think it's too messy to have them both as wrapper classes, so I think we should just call this part of the Envoy file semantics, rather than biasing toward Stats.

And the raw file can be RawFile, whicih will pretty much only be used in the implementation of File, but makes it easier to support multiple OSs.

Reply:
Sounds good to us

Reply:
Big +1 from me.

Reply:
Slightly more thoughts on this: I think it might make sense to have separate decorator classes for file-flushing and stat-collecting, to be determined when constructing the Filesystem that goes into the API, but those don't change the interface; we should still be working with Filesystem everywhere.

Then if we want a new mix-in we can add it, and it can just have one job.

Or I'm fine having them all in Envoy::Filesystem::Instance if @mattklein123 wants to stick with that :)

Reply:
I considered that briefly in the course of this PR, but it looks like it may be tricky pull the stats and the file-flushing apart (e.g. the File object keeps a `flushed_by_timer_` stat)

Reply:
OK simpler to leave them together then. But the API exposed to the rest of the system can be agnostic to that, right?

Reply:
To be sure I understand: you're suggesting `RawFile` and `File` should share a common interface, and the API filesystem will return that inteface (with the implementation being a `File`)?